### PR TITLE
[2024-03-19] heewon #8

### PR DESCRIPTION
**Pull request by HeewonKwak**
## 변수 설명
- `dice_set`: 주사위 쌍 세트 ex) 주사위 4개 경우 : [ [(0, 1), (2, 3)], [(0, 2), (1, 3)], [(0, 3), (1, 2)] ]
- `case_x`: X의 주사위 합의 경우 딕셔너리로 저장 ex) 기본 주사위: {1: 1, 2: 1, 3: 1, 4: 1, 5: 1, 6: 1}
- `win, draw, loss`: 각 경우의 win, draw, loss 수 ex) win : 704, draw: 16, loss: 576

## 접근 방식
- 주사위를 나누는 경우 구하기
- 나눈 주사위들로 나올 수 있는 경우 구하기
- 승률이 가장 높은 경우를 저장
- 모든 경우를 구하지만 시간 복잡도를 최대한 줄이자!
  - `combinations`의 순서가 역순과 쌍을 이루기 때문에 len(dice_case)//2 번만 확인
  - 주사위의 경우를 리스트가 아닌 딕셔너리를 이용하자

## 사용한 모듈
`defaultdict, combinations`

## 추가 정보
- 시간: 1 hour
- 힌트: `None`
- 과거 풀은 경험 有 : 2024 KAKAO WINTER INTERNSHIP 당시에는 Time out!

#8

### CODE REVIEW

File: Programmers/주사위 고르기/heewon.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+from collections import defaultdict
+from itertools import combinations
+
+def solution(dice):
+    answer = []
+    # 승률이 가장 좋은 경우 확인용 변수
+    max_win = 0
+    
+    # 조합으로 구하는 주사위 쌍 구하기
+    dice_case = list(combinations([i for i in range(len(dice))], len(dice)//2))
```
Comment (zsmalla) [inline Programmers/주사위 고르기/heewon.py:13]: `dice_case = list(combinations(range(len(dice)), len(dice)//2))`

동작이 같지만, 불필요한 부분을 생략한 코드입니다!

File: Programmers/주사위 고르기/heewon.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    dice_set = [[dice_case[i], dice_case[-1-i]] for i in range(len(dice_case)//2)]
+
+    for a, b in dice_set:
+        # A, B의 주사위 합의 경우 딕셔너리로 저장
+        case_a = defaultdict(int)
+        case_b = defaultdict(int)
+
+        # A의 주사위 합의 빈도수 구하기
```
Comment (zsmalla) [inline Programmers/주사위 고르기/heewon.py:21]: 주사위 합의 빈도수로 중복되는 합에 대한 계산을 생략하는 아이디어가 좋은 것 같습니다. 자연스레 연산량도 줄어드는 좋은 접근이라고 생각합니다!

**Review by zsmalla — APPROVED** [review]
저는 풀이하면서 중복되는 연산을 묵과하고 진행한 결과 TLE가 났었습니다! 해당풀이는 중복되는 부분을 철저히 배제하신 좋은 풀이라고 생각되네요.

특히 마지막 부분에 loss일 경우에 대한 고려를 한 번 더 해줘서 고려해야 하는 조합을 반으로 줄이신 것이 좋았습니다! 고생하셨어요~

**Review by osmin625 — COMMENTED** [review]
최댓값을 추적하는 방식으로 저와 비슷하게 짜신 것 같아요.
주석을 깔끔하게 작성해주셔서 보기 좋았습니다~

File: Programmers/주사위 고르기/heewon.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                case_b = c_b
+
+        # A, B의 승패 구하기 
+        win, draw, loss = 0, 0, 0
+        for aa in case_a.keys():
+            for bb in case_b.keys():
+                if aa > bb:
+                    win += case_a[aa] * case_b[bb]
+                elif aa == bb:
+                    draw += case_a[aa] * case_b[bb]
```
Comment (limstonestone) [inline Programmers/주사위 고르기/heewon.py:56]: 딱히 의미 없긴하지만 draw 가 없어도 되긴 하겠네요!

**Review by limstonestone — APPROVED** [review]
희원님 저는 이해를 제대로는 못한 것 같아요 ...ㅜㅜㅜ
구체적인 로직은 잘은 이해못했지만 어떤 느낌으로 접근하신지는 알 것 같아요!
중복되는 계산을 줄이는 것이 관건이었는데 딕셔너리를 잘 활용하신 것 같습니다!!
풀이 해설을 참석하지 못해서 아쉽네요 ㅜㅜ 수고하셨습니다!!
